### PR TITLE
Optimize memory resizing logic and prepare it for use in MINIMAL_RUNTIME

### PR DESCRIPTION
**Pull request by juj**

### CODE REVIEW

File: src/preamble.js [hunk 1]
```diff
@@ -297,7 +297,8 @@ var HEAP,
 /** @type {Float64Array} */
   HEAPF64;
 
-function updateGlobalBufferViews() {
+function updateGlobalBufferAndViews(buf) {
+  Module['buffer'] = buffer = buf;
```
Comment (kripken) [inline src/preamble.js]: why is this added? this increases code size unnecessarily I think.

File: src/library.js [hunk 2]
```diff
@@ -631,23 +615,15 @@ LibraryManager.library = {
...
       return false;
     }
 
-#if USE_PTHREADS
-    // Updating the views here is not strictly necessary, since we instrument each load and store
-    // to do so, but doing it here is clean and may be more efficient (avoid surprising the JIT
-    // later by taking a never-taken branch).
-#endif
-    updateGlobalBufferViews();
+    HEAPU32[DYNAMICTOP_PTR>>2] = requestedSize;
```
Comment (kripken) [inline src/library.js]: why is this added? (it doesn't seem moved as I don't see a corresponding deletion?)

File: src/preamble.js (hunk 1 above)
Comment (juj) [reply src/preamble.js]: When I authored this change in our fork, the preamble code used to read
```js
function updateGlobalBuffer(buf) {
  Module['buffer'] = buffer = buf;
}

function updateGlobalBufferViews() {
  Module['HEAP8'] = HEAP8 = new Int8Array(buffer);
  Module['HEAP16'] = HEAP16 = new Int16Array(buffer);
  Module['HEAP32'] = HEAP32 = new Int32Array(buffer);
  Module['HEAPU8'] = HEAPU8 = new Uint8Array(buffer);
  Module['HEAPU16'] = HEAPU16 = new Uint16Array(buffer);
  Module['HEAPU32'] = HEAPU32 = new Uint32Array(buffer);
  Module['HEAPF32'] = HEAPF32 = new Float32Array(buffer);
  Module['HEAPF64'] = HEAPF64 = new Float64Array(buffer);
}
```
so I merged the two functions into one since the two actions always need to be done together.

However it looks like since then the repository has been refactored to remove `updateGlobalBuffer` function and I cannot find an assignment to `Module['buffer'] = buffer = buf;` anywhere after a memory resize occurs? I take this to mean that the field `Module['buffer']` has since been deprecated? I'll remove the assignment.

File: src/library.js (hunk 2 above)
Comment (juj) [reply src/library.js]: Good point. The assignment occurs in sbrk() and brk(), and in our fork I have cleaned up those functions, but did not put the changes in this PR. Dropped this assignment, it's indeed unnecessary.

File: src/library.js [hunk 3]
```diff
@@ -494,44 +494,27 @@ LibraryManager.library = {
...
-      }
-    } catch(e) {
-#if ASSERTIONS
-      console.error('emscripten_realloc_buffer: Attempted to grow from ' + oldSize  + ' bytes to ' + size + ' bytes, but got error: ' + e);
-#endif
-      return false;
-    }
+#if WASM
+      // round size grow request up to wasm page size (fixed 64KB per spec)
+      wasmMemory.grow((size - buffer.byteLength + 65535) >> 16); // .grow() takes a delta compared to the previous size
```
Comment (kripken) [inline src/library.js]: Might be nice to use `{{{ WASM_PAGE_SIZE - 1 }}}` instead of `65535`, and `{{{ Math.log2(WASM_PAGE_SIZE) }}}` instead of `16`. But maybe that doesn't help readability, and this value is not expected to ever change.
